rest-server/src/templates/dockerContainerScript.mustache

```diff
@@ -45,29 +49,43 @@ export PAI_TASK_ROLES_NUM={{{ taskRolesNumber }}}
 export PAI_KILL_ALL_ON_COMPLETED_TASK_NUM={{{ jobData.killAllOnCompletedTaskNumber }}}
 
 export PAI_CURRENT_CONTAINER_PORT=$(( (RANDOM % 55535) + 10001 ))
+# touch a container id file in "TASK_ROLE_NO-TASK_INDEX-CONTAINER_IP-CONTAINER_PORT" format
+# on hdfs to communicate with other containers
 hdfs dfs -touchz $HDFS_LAUNCHER_PREFIX/$PAI_USERNAME/$PAI_JOB_NAME/tmp/$PAI_TASK_ROLE_NO-$PAI_TASK_INDEX-$PAI_CURRENT_CONTAINER_IP-$PAI_CURRENT_CONTAINER_PORT || exit 1
-while [ `hdfs dfs -ls $HDFS_LAUNCHER_PREFIX/$PAI_USERNAME/$PAI_JOB_NAME/tmp/ | wc -l` -lt  $PAI_TASKS_NUM ]; do
+while [ `hdfs dfs -ls $HDFS_LAUNCHER_PREFIX/$PAI_USERNAME/$PAI_JOB_NAME/tmp/ | grep "/$PAI_JOB_NAME/tmp" | wc -l` -lt  $PAI_TASKS_NUM ]; do
+  printf "%s %s\n" "[INFO]" "Waiting for other containers ..."
   sleep 10
 done
 hdfs dfs -ls $HDFS_LAUNCHER_PREFIX/$PAI_USERNAME/$PAI_JOB_NAME/tmp/ \
   | grep "/$PAI_JOB_NAME/tmp" \
   | grep -oE "[^/]+$" \
   | sort -n -k 2 -t"-" \
   > ContainerList
+if [ "$(cat ContainerList | wc -l)" -ne $PAI_TASKS_NUM ]; then
+  printf "%s %s\n%s\n%s\n\n" \
+    "[ERROR]" "ContainerList" \
+    "$(cat ContainerList)" \
+    "$(cat ContainerList | wc -l) containers are available, less than $PAI_TASKS_NUM, exit ..."
+  exit 2
+fi
 
 export PAI_TASK_ROLE_INDEX=$((`cat ContainerList | grep "^$PAI_TASK_ROLE_NO-" | grep -n "$PAI_TASK_ROLE_NO-$PAI_TASK_INDEX-$PAI_CURRENT_CONTAINER_IP-$PAI_CURRENT_CONTAINER_PORT" | cut -d ":" -f 1`-1))
 for i in `seq 0 $((PAI_TASK_ROLES_NUM-1))`; do
   export PAI_TASK_ROLE_${i}_HOST_LIST=`cat ContainerList | grep "^$i-" | cut -d "-" -f 3-4 | tr "-" ":" | sed -e :E -e "N;s/\n/,/;tE"`
 done
 
+printf "%s %s\n%s\n\n" "[INFO]" "ENV" "$(printenv | sort)"
+
 cd $HOME
 hdfs dfs -get $PAI_CODE_DIR code || exit 1
```

Comment:
>hdfs dfs -get $PAI_CODE_DIR code  [](start = 0, length = 33)

by the way, can we leverage launcher to download the code before launching docker instance?

Reply:
I think yes, launcher won't rename it so we have to parse the dir name here if we download it using launcher.

Reply:
let's do this later.

---
In reply to: [162252092](https://github.com/Microsoft/pai/pull/235#discussion_r162252092) [](ancestors = 162252092)